src/libraries/System.Runtime/tests/System.IO.FileSystem.Tests/RandomAccess/NonSeekable.cs

```diff
@@ -0,0 +1,429 @@
+
+                await Assert.ThrowsAsync<UnauthorizedAccessException>(async () => await RandomAccess.ReadAsync(writeHandle, GenerateVectors(1, 1), 0));
+                await Assert.ThrowsAsync<UnauthorizedAccessException>(async () => await RandomAccess.WriteAsync(readHandle, GenerateReadOnlyVectors(1, 1), 0));
+            }
+        }
+
+        [Fact]
+        public async Task ThrowsTaskAlreadyCanceledForCancelledTokenAsync()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                CancellationTokenSource cts = RandomAccess_Base<byte>.GetCancelledTokenSource();
+                CancellationToken token = cts.Token;
+
+                Assert.True(RandomAccess.ReadAsync(readHandle, new byte[1], 0, token).IsCanceled);
+                Assert.True(RandomAccess.WriteAsync(writeHandle, new byte[1], 0, token).IsCanceled);
+                Assert.True(RandomAccess.ReadAsync(readHandle, GenerateVectors(1, 1), 0, token).IsCanceled);
+                Assert.True(RandomAccess.WriteAsync(writeHandle, GenerateReadOnlyVectors(1, 1), 0, token).IsCanceled);
+
+                TaskCanceledException ex = await Assert.ThrowsAsync<TaskCanceledException>(() => RandomAccess.ReadAsync(readHandle, new byte[1], 0, token).AsTask());
+                Assert.Equal(token, ex.CancellationToken);
+                ex = await Assert.ThrowsAsync<TaskCanceledException>(() => RandomAccess.WriteAsync(writeHandle, new byte[1], 0, token).AsTask());
+                Assert.Equal(token, ex.CancellationToken);
+                ex = await Assert.ThrowsAsync<TaskCanceledException>(() => RandomAccess.ReadAsync(writeHandle, GenerateVectors(1, 1), 0, token).AsTask());
+                Assert.Equal(token, ex.CancellationToken);
+                ex = await Assert.ThrowsAsync<TaskCanceledException>(() => RandomAccess.WriteAsync(writeHandle, GenerateReadOnlyVectors(1, 1), 0, token).AsTask());
+                Assert.Equal(token, ex.CancellationToken);
+            }
+        }
+
+        [Fact]
+        public async Task ReadToAnEmptyBufferReturnsZeroWhenDataIsAvailable()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                byte[] content = RandomNumberGenerator.GetBytes(BufferSize);
+                ValueTask write = RandomAccess.WriteAsync(writeHandle, content, fileOffset: 0);
+
+                Assert.Equal(0, RandomAccess.Read(readHandle, Array.Empty<byte>(), fileOffset: 0)); // what we test
+                byte[] buffer = new byte[content.Length * 2];
+
+                ReadExactly(readHandle, buffer, content.Length); // what is required for the above write to succeed
+
+                Assert.Equal(content, buffer.AsSpan(0, content.Length).ToArray());
```

Comment:
I thought we always used `AssertExtensions.SequenceEqual` for array comparisons.